### PR TITLE
Geodes

### PR DESCRIPTION
**Pull request by infiniRekt**

### CODE REVIEW

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeTokenFacet.sol [hunk 1]
```diff
@@ -114,7 +114,7 @@ contract ForgeTokenFacet is Modifiers {
         uint256 amount,
         bytes memory data
     ) public {
-        require(from == LibMeta.msgSender() || isApprovedForAll(from, LibMeta.msgSender()), "ForgeTokenFacet: caller is not token owner or approved");
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeTokenFacet.sol:117]: Why was this change necessary? @infiniRekt

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol [hunk 2]
```diff
@@ -0,0 +1,249 @@
+import "../libraries/LibAppStorage.sol";
+
+import {LibMeta} from "../../../shared/libraries/LibMeta.sol";
+import {ILink} from "../../interfaces/ILink.sol";
+import {ForgeFacet} from "./ForgeFacet.sol";
+import {ForgeTokenFacet} from "./ForgeTokenFacet.sol";
+import {LibSvg} from "../../libraries/LibSvg.sol";
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:7]: Don't think this import is needed @infiniRekt

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol [hunk 3]
```diff
@@ -0,0 +1,249 @@
+import "../libraries/LibAppStorage.sol";
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:1]: No pragma solidity version, need to add

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+    function keyHash() external view returns (bytes32) {
+        return s.keyHash;
+    }
+
+    function getMaxVrf() external view returns (uint256) {
+        return MAX_VRF;
+    }
+
+    function areGeodePrizesAvailable() public view returns (bool) {
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:46]: This function is called in openGeodes, which specifies geodeTokenIds. It might be better to also pass through a tokenIds arg to this function, otherwise I think it may return false for some situations where geode prizes are unavailable for some tokenIds but not others.

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+        VrfRequestInfo storage info = s.vrfRequestIdToVrfRequestInfo[requestId];
+
+        require(info.status == VrfStatus.READY_TO_CLAIM, "ForgeVRFFacet: not ready to claim");
+        require(info.randomNumber != 0, "ForgeVRFFacet: invalid random number");
+
+        uint256 modNum = 10000;
+        uint256 divNum = 1;
+
+        uint256 numTotalPrizesLeft = numTotalPrizesLeft();
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:184]: You shouldn't declare a variable with the same name as a function @infiniRekt

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        for (uint256 i; i < _amountPerToken.length; i++) {
+            total += _amountPerToken[i];
+
+            require(_geodeTokenIds[i] >= GEODE_COMMON && _geodeTokenIds[i] <= GEODE_GODLIKE, "ForgeVRFFacet: Invalid geode token ID");
+            require(forgeTokenFacet().balanceOf(sender, _geodeTokenIds[i]) >= _amountPerToken[i], "ForgeVRFFacet: not enough geodes owned");
+        }
+
+        // spend geodes
+        for (uint256 i; i < _geodeTokenIds.length; i++) {
+            forgeTokenFacet().safeTransferFrom(sender, address(this), _geodeTokenIds[i], _amountPerToken[i], "");
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:84]: The geodes don't get burned? Just sent to the forge diamond? @infiniRekt

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        uint256 total;
+        for (uint256 i; i < _amountPerToken.length; i++) {
+            total += _amountPerToken[i];
+
+            require(_geodeTokenIds[i] >= GEODE_COMMON && _geodeTokenIds[i] <= GEODE_GODLIKE, "ForgeVRFFacet: Invalid geode token ID");
+            require(forgeTokenFacet().balanceOf(sender, _geodeTokenIds[i]) >= _amountPerToken[i], "ForgeVRFFacet: not enough geodes owned");
+        }
+
+        // spend geodes
+        for (uint256 i; i < _geodeTokenIds.length; i++) {
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:83]: Could this code be included in the same for loop above? @infiniRekt

Comment (cinnabarhorse) [reply contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:83]: Or perhaps you could use the `safeBatchTransferFrom` function instead of looping @infiniRekt

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol (hunk 6 above)
Comment (cinnabarhorse) [reply contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:84]: Ah I see, because they are refunded if there are no prizes left

Comment (cinnabarhorse) [reply contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:84]: In that case maybe you would want to burn the geode in the claimWinnings function?

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeDAOFacet.sol [hunk 8]
```diff
@@ -133,35 +135,47 @@ contract ForgeDAOFacet is Modifiers {
...
+        s.geodeWinChanceBips[MYTHICAL_RSM] = chances.mythical;
+        s.geodeWinChanceBips[GODLIKE_RSM] = chances.godlike;
+
+        emit SetGeodeWinChance(chances);
+    }
+
+    // @notice Allow DAO to set which prizes can be won from a Geode.
+    // @param ids Token IDs of the available prizes
+    // @param quantities Initial amounts of each prize available
+    function setGeodePrizes(uint256[] calldata ids, uint256[] calldata quantities) external onlyDaoOrOwner {
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeDAOFacet.sol:154]: Shouldn't this function also transfer prizes from msg.sender into the Forge diamond? @infiniRekt

Comment (cinnabarhorse) [reply contracts/Aavegotchi/ForgeDiamond/facets/ForgeDAOFacet.sol:154]: That's what we do with the raffle contract

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeDAOFacet.sol [hunk 9]
```diff
@@ -133,35 +135,47 @@ contract ForgeDAOFacet is Modifiers {
...
+            s.geodePrizeQuantities[ids[i]] = quantities[i];
+        }
+
+        emit SetGeodePrizes(ids, quantities);
+    }
+
+    function getGeodePrizesRemaining() external view returns (uint256[] memory, uint256[] memory) {
+        uint256[] memory quantities = new uint256[](s.geodePrizeTokenIds.length);
+        for (uint256 i; i < s.geodePrizeTokenIds.length; i++) {
+            quantities[i] = s.geodePrizeQuantities[s.geodePrizeTokenIds[i]];
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeDAOFacet.sol:175]: I'm not totally sure you need the `geodePrizeQuantities` storage variable. You could just use the `geodePrizeTokenIds` array and then track the balance of each tokenId in the contract. 

The issue with using "virtual" supply is that there could be discrepancies between the supply that the contract thinks it has (geodePrizeQuantities) and the amount actually in the contract (the token balance). Would be cleaner to use the token balance itself I think

Comment (infiniRekt) [reply contracts/Aavegotchi/ForgeDiamond/facets/ForgeDAOFacet.sol:175]: I prefer to use a "virtual" supply to keep tighter control over what is a geode prize instead of whatever is sent to the contract.

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeTokenFacet.sol (hunk 1 above)
Comment (infiniRekt) [reply contracts/Aavegotchi/ForgeDiamond/facets/ForgeTokenFacet.sol:117]: transfers were reverting when called from the vrf facet.

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol (hunk 4 above)
Comment (infiniRekt) [reply contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:46]: this function and its usage is just meant to check whether there are any prizes left at all, not for any specific ids.

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        require(_geodeTokenIds.length == _amountPerToken.length, "ForgeVRFFacet: mismatched arrays");
+
+        uint256 totalGeodes;
+        for (uint256 i; i < _amountPerToken.length; i++) {
+            totalGeodes += _amountPerToken[i];
+        }
+        require(totalGeodes <= MAX_VRF, "ForgeVRFFacet: Exceeds max total geodes per call");
+
+        address sender = LibMeta.msgSender();
+        uint256 total;
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:76]: Don't think this variable is being used anymore @infiniRekt

File: contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        VrfRequestInfo storage info = s.vrfRequestIdToVrfRequestInfo[requestId];
+
+        require(info.status == VrfStatus.READY_TO_CLAIM, "ForgeVRFFacet: not ready to claim");
+        require(info.randomNumber != 0, "ForgeVRFFacet: invalid random number");
+
+        uint256 modNum = 10000;
+        uint256 divNum = 1;
+
+        uint256 prizesLeft = numTotalPrizesLeft();
+        uint256 numWins;
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/ForgeDiamond/facets/ForgeVRFFacet.sol:185]: Should `numWins` get reset when you set geode prizes? If it keeps incrementing, there could eventually be a situation where it gets higher than the `prizesLeft`, even though you just reset with new prizes. @infiniRekt
